### ISSUE TITLE
StructurePool constructor will lag and eat RAM like crazy if the element's weight is set to an extreme value

### ISSUE THREAD

**Issue by TelepathicGrunt** [jellysquid3/lithium-fabric#157]
So for jigsaw structures such as villages or outposts, they have what are called pools and in these pools are pieces of the structure with a weight of how often they get picked. What has been reported is that if someone sets a piece's weight to a very high value by datapack or a mod, the game will lag hard when loading up the world and when going to the structure. If the weight is high enough, the JVM will run out of memory!

Now if you check out StructurePool's constructor, you'll see why very fast...
https://hatebin.com/ylpqtmsqkg

Yep. Mojang makes a new instance of the element x number of times with x being the weight and shoves all those separate instances into a list just so they can do a `this.elements.get(random.nextInt(this.elements.size()));` to get a random piece. It's uber cursed lmao 

Here's a datapack to reproduce the issue!
[outofmemory.zip](https://github.com/jellysquid3/lithium-fabric/files/5461353/outofmemory.zip)

And here's the result of the datapack 
![image](https://user-images.githubusercontent.com/40846040/97633792-07a2bc00-1a0b-11eb-85de-a7f8bd6895e1.png)

**Comment by TelepathicGrunt** [comment]
made mojang issue report as well so hopefully mojang will fix it in the future. https://bugs.mojang.com/browse/MC-203131

**Comment by maityyy** [comment]
I don't think it's possible to fix it in Lithium in any way. More suitable for Hydrogen... Maybe we should be lazy to initialize this list?

**Comment by Kroppeb** [comment]
> I don't think it's possible to fix it in Lithium in any way. More suitable for Hydrogen... Maybe we should be lazy to initialize this list?

This is actually quite easy to fix. IIRC Minecraft even has a class already for something like this that does not use all the memory.

**Comment by maityyy** [comment]
> > I don't think it's possible to fix it in Lithium in any way. More suitable for Hydrogen... Maybe we should be lazy to initialize this list?
> 
> This is actually quite easy to fix. IIRC Minecraft even has a class already for something like this that does not use all the memory.

You write about `net.minecraft.util.Lazy`?

**Comment by Kroppeb** [comment]
> > > I don't think it's possible to fix it in Lithium in any way. More suitable for Hydrogen... Maybe we should be lazy to initialize this list?
> > 
> > 
> > This is actually quite easy to fix. IIRC Minecraft even has a class already for something like this that does not use all the memory.
> 
> You write about `net.minecraft.util.Lazy`?

no, cause it would be initialized anyway. I meant `WeightedPicker`
